qemu/tests/pci_bridge_hotplug_unplug.py

```diff
@@ -0,0 +1,163 @@
+    image_params["drive_format"] = drive_format
+    devices = []
+
+    if drive_format == 'usb2':
+        usbc_params = {'usb_type': 'usb-ehci'}
+        devices = vm.devices.usbc_by_params(drive_format,
+                                            usbc_params,
+                                            pci_bus={'aobject': parent_bus})
+
+    devices += vm.devices.images_define_by_params(image, image_params, 'disk',
+                                                  None, False, None,
+                                                  pci_bus={'aobject': parent_bus})
+
+    for device in devices:
+        vm.devices.insert(device)
+        device.hotplug(vm.monitor)
+
+    ver_out = devices[-1].verify_hotplug("", vm.monitor)
+    if not ver_out:
+        test.fail("%s disk is not in qtree after hotplug!" % drive_format)
+
+    if drive_format == 'usb2':
+        usb_serial = params["blk_extra_params_%s" % image].split("=")[1]
+        res = utils_misc.wait_for(check_usb_in_guest, timeout,
+                                  text="Wait for getting usb device info")
+        if res is None:
+            test.fail("Could not find the usb device serial:[%s]" % usb_serial)
+
+    if drive_format == 'virtio':
+        return [devices[-1]]
+    else:
+        return devices[::2]
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    [pci-bridge] Hotplug and unplug different devices to 1 pci-bridge
+    [pci-bridge] Hotplug and unplug different devices to different pci-bridges
+    This case will:
+    1) Start the guest.
+    2) Hotplug different devices to pci-bridge(s).
+    3) Check devices.
+    4) Hot-unplug devices.
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+    pci_bridge_num = int(params.get("pci_bridge_num", 1))
+    if pci_bridge_num > 1:
+        raise NotImplementedError()
+
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+
+    login_timeout = int(params.get("login_timeout", 360))
+    session = vm.wait_for_login(timeout=login_timeout)
+    cmd_timeout = int(params.get("cmd_timeout", 3600))
+
+    image_list = params.objects("images")
+    image_num = len(image_list)
+    if (image_num < 2):
+        test.cancel("No pre-prepared images for hotplugging!")
+
+    device_list = []
+    fmt_list = params["disk_driver"].split()
+    if pci_bridge_num < 1:
+        test.cancel("There should be at least one pci-bridge!")
+    elif pci_bridge_num == 1:
+        pci_bridge_id = params["pci_controllers"]
+        for i in range(image_num - 1):
+            if i >= len(fmt_list):
+                test.error("No sufficient disk driver type!")
+            error_context.context("Hotplug a %s disk!" % fmt_list[i],
+                                  logging.info)
+            device_list += disk_hotplug(test, params, vm, session,
+                                        image_list.pop(), fmt_list[i],
```

Comment:
we can use `image_list[i+1]` instead of `image_list.pop()`. In this way, only get `image_list `once is enough.